code/renderers/preact/src/public-types.ts

```diff
@@ -19,21 +23,54 @@ export type { PreactRenderer };
  *
  * @see [Default export](https://storybook.js.org/docs/formats/component-story-format/#default-export)
  */
-export type Meta<TArgs = Args> = ComponentAnnotations<PreactRenderer, TArgs>;
+export type Meta<TCmpOrArgs = Args> = [TCmpOrArgs] extends [ComponentType<any>]
+  ? ComponentAnnotations<PreactRenderer, ComponentProps<TCmpOrArgs>>
+  : ComponentAnnotations<PreactRenderer, TCmpOrArgs>;
 
 /**
  * Story function that represents a CSFv2 component example.
  *
  * @see [Named Story exports](https://storybook.js.org/docs/formats/component-story-format/#named-story-exports)
  */
-export type StoryFn<TArgs = Args> = AnnotatedStoryFn<PreactRenderer, TArgs>;
+export type StoryFn<TCmpOrArgs = Args> = [TCmpOrArgs] extends [ComponentType<any>]
+  ? AnnotatedStoryFn<PreactRenderer, ComponentProps<TCmpOrArgs>>
+  : AnnotatedStoryFn<PreactRenderer, TCmpOrArgs>;
 
 /**
  * Story object that represents a CSFv3 component example.
  *
  * @see [Named Story exports](https://storybook.js.org/docs/formats/component-story-format/#named-story-exports)
  */
-export type StoryObj<TArgs = Args> = StoryAnnotations<PreactRenderer, TArgs>;
+export type StoryObj<TMetaOrCmpOrArgs = Args> = [TMetaOrCmpOrArgs] extends [
+  {
+    render?: ArgsStoryFn<PreactRenderer, any>;
+    component?: infer Component;
+    args?: infer DefaultArgs;
+  },
+]
+  ? Simplify<
+      (Component extends ComponentType<any> ? ComponentProps<Component> : unknown) &
+        ArgsFromMeta<PreactRenderer, TMetaOrCmpOrArgs>
+    > extends infer TArgs
+    ? StoryAnnotations<
+        PreactRenderer,
+        AddMocks<TArgs, DefaultArgs>,
+        SetOptional<TArgs, keyof TArgs & keyof DefaultArgs>
+      >
+    : never
+  : TMetaOrCmpOrArgs extends ComponentType<any>
+    ? StoryAnnotations<PreactRenderer, ComponentProps<TMetaOrCmpOrArgs>>
+    : StoryAnnotations<PreactRenderer, TMetaOrCmpOrArgs>;
+
+// This performs a downcast to function types that are mocks, when a mock fn is given to meta args.
+type AddMocks<TArgs, DefaultArgs> = Simplify<{
+  [T in keyof TArgs]: T extends keyof DefaultArgs
+    ? // eslint-disable-next-line @typescript-eslint/ban-types
+      DefaultArgs[T] extends (...args: any) => any & { mock: {} } // allow any function with a mock object
```

Comment:
Can you explain out the `mock` use case?

Reply:
I pretty much just copied React's types here. I assume this is for `@storybook/test`'s `fn()` result.